File: champ_dhonneur/ia/encodage.py
"""Encodage des positions et des actions pour le réseau de neurones (2 joueurs).

Principes :
  * **Point de vue du joueur au trait.** Tout est exprimé en « moi / adversaire ».
    Le plateau 2J est symétrique par rotation de 180° : pour Noir, on tourne
    le plateau, si bien que le réseau voit toujours sa propre armée en bas.
  * **Aucune fuite d'information.** On n'encode que ce que le joueur au trait
    sait réellement : sa main, la composition de son sac, sa défausse cachée ;
    pour l'adversaire, seulement le multiensemble {sac + main + défausse cachée}
    (déductible de l'information publique) et la taille de chaque zone.
  * **Actions par caractéristiques.** Au lieu d'un immense vecteur de sortie
    fixe, chaque action légale est décrite par quelques entiers (type, pièce,
    unité, cases). Le réseau calcule un score par action à partir de ces
    entiers et des représentations des cases concernées (réseau « pointeur »).
    Cela généralise naturellement aux 1820 compositions d'armées possibles.
  * **Mise en place avancée (draft).** Les 8 jetons d'unité sont les 8 cartes en jeu :
    mes unités (camp 0), celles de l'adversaire (camp 1), puis les cartes encore
    disponibles (camp 2). Choisir une carte est une action comme les autres. Hors draft,
    les caractéristiques propres au draft sont nulles : une position issue d'un draft
    s'encode exactement comme la même position à armées imposées.

Version 2 de l'encodage (`VERSION`) : incompatible avec les modèles de la version 1.

Tout est en numpy pur pour pouvoir encoder dans les processus d'auto-jeu.
"""
from __future__ import annotations

import numpy as np

from ..board import get_board
from ..engine import (ATTACK, BOLSTER, CONTROL, DEPLOY, DRAFT, INITIATIVE, MOVE, PASS, RECRUIT,
                      RG_RESERVE, RG_UNIT, SKIP, TACTIC, Action, Game)
from ..units import ALL_LETTERS, ROYAL, UNITS

VERSION = 2

SPEC = get_board("2J")
N_CELLS = SPEC.n_cells                 # 37
NONE_CELL = N_CELLS                    # index « pas de case »
N_UNIT_TOKENS = 8                      # les 8 cartes en jeu : à moi, adverses, disponibles
N_TOKENS = N_CELLS + N_UNIT_TOKENS + 1  # + jeton global

# types de pièces : 0 = aucune, 1..16 = unités, 17 = Sceau royal
COIN_ID = {u: i + 1 for i, u in enumerate(ALL_LETTERS)}
COIN_ID[ROYAL] = len(ALL_LETTERS) + 1
N_COIN_TYPES = len(ALL_LETTERS) + 2    # 18

KINDS = [DEPLOY, BOLSTER, MOVE, CONTROL, ATTACK, TACTIC, INITIATIVE, RECRUIT, PASS,
         SKIP, RG_RESERVE, RG_UNIT, DRAFT]
KIND_ID = {k: i for i, k in enumerate(KINDS)}
N_KINDS = len(KINDS)

PENDING_KINDS = ["", "berserk", "soldat", "merc", "footman", "priest", "rg", "draft"]
PENDING_ID = {k: i for i, k in enumerate(PENDING_KINDS)}

MAX_COINS_EMB = 8

# Rotation de 180° : (colonne x, rang n) -> (6 - x, longueur - n + 1). Involution.
_ROT = [SPEC.index[f"{'abcdefg'[6 - x]}{SPEC.col_lengths[6 - x] - n + 1}"] for x, n in SPEC.cells]
PERM = [list(range(N_CELLS)), _ROT]    # PERM[équipe][case réelle] = case vue
LOC_SET = frozenset(SPEC.locations)

CELL_F = 3    # flotteurs par case
UNIT_F = 12   # flotteurs par jeton d'unité
GLOB_F = 26   # flotteurs du jeton global
ACT_F = 7     # entiers par action : type, pièce, unité, extra, c0, c1, c2


def encode_state(g: Game) -> dict[str, np.ndarray]:
    """Observation du joueur au trait (le jeu doit être en mode 2J)."""
    p = g.to_move
    me, opp = g.players[p], g.players[1 - p]
    t = me.team
    perm = PERM[t]

    cell_i = np.zeros((N_CELLS, 4), np.int64)   # unité, propriétaire, contrôle, pièces
    cell_f = np.zeros((N_CELLS, CELL_F), np.float32)
    for loc in SPEC.locations:
        c = g.control[loc]
        cell_i[perm[loc], 2] = 1 if c is None else (2 if c == t else 3)
    board_coins = [{}, {}]
    n_units = [{}, {}]
    for pos, u in g.board.items():
        v = perm[pos]
        side = 0 if u.owner == p else 1
        cell_i[v, 0] = COIN_ID[u.utype]
        cell_i[v, 1] = 1 + side
        cell_i[v, 3] = min(u.coins, MAX_COINS_EMB)
        cell_f[v, 0] = u.coins / 4.0
        board_coins[side][u.utype] = board_coins[side].get(u.utype, 0) + u.coins
        n_units[side][u.utype] = n_units[side].get(u.utype, 0) + 1
    pending = g.pending[-1] if g.pending else None
    if pending is not None:
        if pending.pos >= 0:
            cell_f[perm[pending.pos], 1] = 1.0
        for pos in pending.positions:
            cell_f[perm[pos], 2] = 1.0

    ui = []
    uf = []
    dr = g.draft if g.in_draft else None
    for side, pl in enumerate((me, opp)):
        bc, nu = board_coins[side], n_units[side]
        hand, bag, dd, du = pl.hand, pl.bag, pl.disc_down, pl.disc_up
        for u in sorted(pl.units):
            ui.append((COIN_ID[u], side))
            h, b, d, up = hand.count(u), bag.count(u), dd.count(u), du.count(u)
            mine = side == 0
            uf.append((h / 2.0 if mine else 0.0, b / 2.0 if mine else 0.0, d / 2.0 if mine else 0.0,
                       (h + b + d) / 3.0, up / 2.0, pl.reserve.get(u, 0) / 3.0,
                       pl.lost.get(u, 0) / 3.0, bc.get(u, 0) / 3.0, float(nu.get(u, 0)),
                       UNITS[u].count / 5.0, (h + b + d + up) / 4.0, 1.0 if (mine and h) else 0.0))
    if dr is not None:
        for u in dr.available:
            ui.append((COIN_ID[u], 2))
            uf.append((0.0,) * 9 + (UNITS[u].count / 5.0, 0.0, 0.0))
    unit_i = np.array(ui, np.int64)
    unit_f = np.array(uf, np.float32)

    glob_i = np.zeros(2, np.int64)
    glob_f = np.zeros(GLOB_F, np.float32)
    if pending is not None:
        glob_i[0] = PENDING_ID[pending.kind]
        if pending.coin is not None:
            glob_i[1] = COIN_ID[pending.coin]
    elif dr is not None:
        glob_i[0] = PENDING_ID["draft"]
    total_markers = 6
    gf = glob_f
    gf[0] = g.markers_left[t] / total_markers
    gf[1] = g.markers_left[1 - t] / total_markers
    gf[2] = 1.0 if g.initiative == p else 0.0
    gf[3] = 1.0 if g.initiative_moved else 0.0
    gf[4] = 1.0 if g.round_first == p else 0.0
    gf[5] = 1.0 if g.can_take_initiative(p) else 0.0
    gf[6] = len(me.hand) / 3.0
    gf[7] = len(opp.hand) / 3.0
    gf[8] = len(me.bag) / 10.0
    gf[9] = len(opp.bag) / 10.0
    gf[10] = (len(me.disc_up) + len(me.disc_down)) / 10.0
    gf[11] = len(opp.disc_down) / 5.0
    gf[12] = len(opp.disc_up) / 10.0
    gf[13] = 1.0 if ROYAL in me.hand else 0.0
    gf[14] = 1.0 if ROYAL in me.bag else 0.0
    gf[15] = 1.0 if (ROYAL in me.disc_down or ROYAL in me.disc_up) else 0.0
    gf[16] = 1.0 if ROYAL in (opp.bag + opp.hand + opp.disc_down) else 0.0
    gf[17] = 1.0 if ROYAL in opp.disc_up else 0.0
    gf[18] = min(g.round, 100) / 50.0
    gf[19] = g.round / g.max_rounds
    gf[20] = 1.0 if g.current == p else 0.0
    gf[21] = sum(1 for loc in SPEC.locations if g.control[loc] is None) / 10.0
    if dr is not None:
        gf[22] = 1.0
        gf[23] = dr.step / len(dr.pool)
        gf[24] = draft_picks_left(g) / 2.0
    gf[25] = 1.0 if g.first_player == p else 0.0
    return {"cell_i": cell_i, "cell_f": cell_f, "unit_i": unit_i, "unit_f": unit_f,
            "glob_i": glob_i, "glob_f": glob_f}


def draft_picks_left(g: Game) -> int:
    """Cartes que le joueur au trait choisit encore dans ce tour de draft (1 ou 2)."""
    from ..units import DRAFT_ORDER
    order, s = DRAFT_ORDER[g.mode], g.draft.step
    n = 1
    while s + n < len(order) and order[s + n] == order[s]:
        n += 1
    return n


def encode_action(a: Action, perm: list[int]) -> tuple[int, ...]:
    cells = [perm[c] for c in a.cells[:3]]
    cells += [NONE_CELL] * (3 - len(cells))
    return (KIND_ID[a.kind],
            COIN_ID[a.coin] if a.coin else 0,
            COIN_ID[a.unit] if a.unit else 0,
            COIN_ID[a.extra] if a.extra else 0,
            *cells)


def encode_actions(g: Game, legal: list[Action]) -> np.ndarray:
    perm = PERM[g.team(g.to_move)]
    return np.array([encode_action(a, perm) for a in legal], np.int64).reshape(-1, ACT_F)


STATE_KEYS = ("cell_i", "cell_f", "unit_i", "unit_f", "glob_i", "glob_f")


def collate(states: list[dict], actions: list[np.ndarray], a_max: int | None = None):
    """Empile des observations et des listes d'actions de tailles variables.

    Renvoie (dict d'entrées numpy, masque (B, A)). Les actions manquantes
    sont remplies avec des zéros et masquées.
    """
    b = len(states)
    out = {k: np.stack([s[k] for s in states]) for k in STATE_KEYS}
    a_len = max(len(a) for a in actions)
    if a_max is not None:
        a_len = max(a_len, a_max)
    acts = np.zeros((b, a_len, ACT_F), np.int64)
    acts[:, :, 4:] = NONE_CELL
    mask = np.zeros((b, a_len), bool)
    for i, a in enumerate(actions):
        acts[i, :len(a)] = a
        mask[i, :len(a)] = True
    out["acts"] = acts
    return out, mask

File: rust/src/encodage.rs
//! Encodage des positions et des actions pour le réseau, identique à `ia/encodage.py`.

use crate::moteur::*;
use crate::plateau::{plateau, AUCUNE_CASE, N_CASES};

pub const CELL_I: usize = 4;
pub const CELL_F: usize = 3;
pub const UNIT_I: usize = 2;
pub const UNIT_F: usize = 12;
pub const GLOB_I: usize = 2;
pub const GLOB_F: usize = 26;
pub const ACT_F: usize = 7;
pub const N_UNITES: usize = 8;
/// Type d'action « choisir une carte » et type de décision « draft » (glob_i[0]) de l'encodage.
pub const ATT_DRAFT: i64 = 7;

/// Observation du joueur au trait, en entiers (i64) et flottants (f32), à plat.
#[derive(Clone)]
pub struct Observation {
    pub cell_i: [i64; N_CASES * CELL_I],
    pub cell_f: [f32; N_CASES * CELL_F],
    pub unit_i: [i64; N_UNITES * UNIT_I],
    pub unit_f: [f32; N_UNITES * UNIT_F],
    pub glob_i: [i64; GLOB_I],
    pub glob_f: [f32; GLOB_F],
}

#[inline]
fn vue(equipe: u8, c: u8) -> usize {
    if equipe == 0 {
        c as usize
    } else {
        plateau().rotation[c as usize] as usize
    }
}

pub fn encoder_etat(g: &Partie) -> Observation {
    let pl = plateau();
    let p = g.au_trait();
    let me = &g.e.joueurs[p as usize];
    let opp = &g.e.joueurs[1 - p as usize];
    let t = me.equipe;
    let mut o = Observation {
        cell_i: [0; N_CASES * CELL_I],
        cell_f: [0.0; N_CASES * CELL_F],
        unit_i: [0; N_UNITES * UNIT_I],
        unit_f: [0.0; N_UNITES * UNIT_F],
        glob_i: [0; GLOB_I],
        glob_f: [0.0; GLOB_F],
    };
    for &l in &pl.lieux {
        let c = g.e.controle[l as usize];
        o.cell_i[vue(t, l) * CELL_I + 2] = if c < 0 { 1 } else if c as u8 == t { 2 } else { 3 };
    }
    let mut pieces_plateau = [[0i32; N_TYPES]; 2];
    let mut n_unites = [[0i32; N_TYPES]; 2];
    for &pos in g.e.ordre.as_slice() {
        let u = g.e.cases[pos as usize];
        let v = vue(t, pos);
        let cote = if u.proprio == p { 0 } else { 1 };
        o.cell_i[v * CELL_I] = u.genre as i64;
        o.cell_i[v * CELL_I + 1] = 1 + cote as i64;
        o.cell_i[v * CELL_I + 3] = (u.pieces as i64).min(8);
        o.cell_f[v * CELL_F] = u.pieces as f32 / 4.0;
        pieces_plateau[cote][u.genre as usize] += u.pieces as i32;
        n_unites[cote][u.genre as usize] += 1;
    }
    let attente = g.attente().copied();
    if let Some(pd) = attente {
        if pd.pos >= 0 {
            o.cell_f[vue(t, pd.pos as u8) * CELL_F + 1] = 1.0;
        }
        for &pos in pd.positions() {
            o.cell_f[vue(t, pos) * CELL_F + 2] = 1.0;
        }
    }
    let mut k = 0;
    for (cote, j) in [me, opp].into_iter().enumerate() {
        let mut unites = j.unites;
        let n = j.n_unites as usize;
        unites[..n].sort();
        let mine = cote == 0;
        for &u in &unites[..n] {
            o.unit_i[k * UNIT_I] = u as i64;
            o.unit_i[k * UNIT_I + 1] = cote as i64;
            let h = j.main.count(u) as f32;
            let b = j.sac.count(u) as f32;
            let d = j.def_cachee.count(u) as f32;
            let up = j.def_visible.count(u) as f32;
            let f = &mut o.unit_f[k * UNIT_F..(k + 1) * UNIT_F];
            f[0] = if mine { h / 2.0 } else { 0.0 };
            f[1] = if mine { b / 2.0 } else { 0.0 };
            f[2] = if mine { d / 2.0 } else { 0.0 };
            f[3] = (h + b + d) / 3.0;
            f[4] = up / 2.0;
            f[5] = j.reserve[u as usize] as f32 / 3.0;
            f[6] = j.perdues[u as usize] as f32 / 3.0;
            f[7] = pieces_plateau[cote][u as usize] as f32 / 3.0;
            f[8] = n_unites[cote][u as usize] as f32;
            f[9] = NOMBRE[u as usize] as f32 / 5.0;
            f[10] = (h + b + d + up) / 4.0;
            f[11] = if mine && h > 0.0 { 1.0 } else { 0.0 };
            k += 1;
        }
    }
    if g.e.en_draft {
        for u in 1..=16u8 {
            if g.e.dispo & (1 << (u - 1)) != 0 {
                o.unit_i[k * UNIT_I] = u as i64;
                o.unit_i[k * UNIT_I + 1] = 2;
                o.unit_f[k * UNIT_F + 9] = NOMBRE[u as usize] as f32 / 5.0;
                k += 1;
            }
        }
    }
    if let Some(pd) = attente {
        o.glob_i[0] = pd.genre as i64;
        o.glob_i[1] = pd.piece as i64;
    } else if g.e.en_draft {
        o.glob_i[0] = ATT_DRAFT;
    }
    let gf = &mut o.glob_f;
    let un = |b: bool| if b { 1.0f32 } else { 0.0 };
    gf[0] = g.e.marqueurs[t as usize] as f32 / 6.0;
    gf[1] = g.e.marqueurs[1 - t as usize] as f32 / 6.0;
    gf[2] = un(g.e.initiative == p);
    gf[3] = un(g.e.init_bougee);
    gf[4] = un(g.e.premier_manche == p);
    gf[5] = un(g.peut_prendre_initiative(p));
    gf[6] = me.main.len() as f32 / 3.0;
    gf[7] = opp.main.len() as f32 / 3.0;
    gf[8] = me.sac.len() as f32 / 10.0;
    gf[9] = opp.sac.len() as f32 / 10.0;
    gf[10] = (me.def_visible.len() + me.def_cachee.len()) as f32 / 10.0;
    gf[11] = opp.def_cachee.len() as f32 / 5.0;
    gf[12] = opp.def_visible.len() as f32 / 10.0;
    gf[13] = un(me.main.contains(ROYAL));
    gf[14] = un(me.sac.contains(ROYAL));
    gf[15] = un(me.def_cachee.contains(ROYAL) || me.def_visible.contains(ROYAL));
    gf[16] = un(opp.sac.contains(ROYAL) || opp.main.contains(ROYAL) || opp.def_cachee.contains(ROYAL));
    gf[17] = un(opp.def_visible.contains(ROYAL));
    gf[18] = (g.e.manche.min(100)) as f32 / 50.0;
    gf[19] = g.e.manche as f32 / g.e.max_manches as f32;
    gf[20] = un(g.e.courant == p);
    gf[21] = pl.lieux.iter().filter(|&&l| g.e.controle[l as usize] < 0).count() as f32 / 10.0;
    if g.e.en_draft {
        let s = g.e.etape_draft as usize;
        let mut n = 1;
        while s + n < N_CARTES && ORDRE_DRAFT[s + n] == ORDRE_DRAFT[s] {
            n += 1;
        }
        gf[22] = 1.0;
        gf[23] = s as f32 / N_CARTES as f32;
        gf[24] = n as f32 / 2.0;
    }
    gf[25] = un(g.e.premier == p);
    o
}

/// Écrit les 7 entiers de chaque action (type, pièce, unité, extra, 3 cases vues).
pub fn encoder_actions(g: &Partie, legal: &[Action], out: &mut Vec<i64>) {
    let t = g.equipe(g.au_trait());
    for a in legal {
        out.push(a.genre as i64);
        out.push(a.piece as i64);
        out.push(a.unite as i64);
        out.push(a.extra as i64);
        for k in 0..3 {
            out.push(if k < a.nc as usize { vue(t, a.cases[k]) as i64 } else { AUCUNE_CASE as i64 });
        }
    }
}
